### DISCUSSION TITLE
EMQX Not able to access on TLS using AWS PrivateLink

### DISCUSSION THREAD

**Discussion by satyamtpatel** [emqx/emqx#10364] [General]
Hi,

We are trying to setup AWS Private Link from VPC A to consume mqtt service installed in VPC B (Provider). We setup NLB with 2 listeners 1) For Application 2) For MQTT over TLS. We are able to reach application over https but for MQTT we are not able to reach using mqtts. When we change NLB target group to TCP then it is allowing both mqtts and mqtt. Ideally, we would like to access mqtts over tsl. What should we do?

**Comment by ieQu1** [comment]
Hello,

1) Can you make sure NLB can access the mqtts port on the broker?
2) "we are not able to reach using mqtts" could you describe symptoms in more detail? Does the client time out, or does the TLS connection terminate immediately?
3) Is there any log on the EMQX broker side when the client tries to connect over mqtts via NLB?

**Comment by imvikysharma** [comment]
It sounds like you have set up an NLB in VPC B to handle incoming MQTT traffic over TLS, and you are trying to access this service from VPC A over a PrivateLink connection.

If you are able to access the MQTT service over TCP using the NLB target group, but not over TLS using the mqtts listener, there may be an issue with the SSL/TLS certificate configuration on the NLB.

Here are some troubleshooting steps you can try:

Verify that the SSL/TLS certificate on the NLB is valid and issued by a trusted certificate authority. You can check the certificate details by accessing the NLB listener in the AWS Management Console.

Check that the MQTT client in VPC A is configured to trust the SSL/TLS certificate of the NLB. The client may be rejecting the certificate because it is not recognized as valid.

Check that the security groups and network ACLs are configured correctly to allow traffic between VPC A and VPC B over the PrivateLink connection.

Verify that the NLB is configured to use the correct SSL/TLS protocol and cipher suite for MQTT traffic. You can check the NLB listener configuration to see if it is using the recommended settings for MQTT over TLS.

**Reply by zmstone** [reply]
> It sounds like you have set up an NLB in VPC B to handle incoming MQTT traffic over TLS, and you are trying to access this service from VPC A over a PrivateLink connection.

@imvikysharma  thank you for your participation, but the answer ~~is not quite relevant~~ seems too generic. Are you AI?

**Comment by zmstone** [comment]
@satyamtpatel 

> When we change NLB target group to TCP

Do you mean:
- if the NLB's target group is configured https, then it does not work
- if the NLB's target group is configured tcp, then it started working

It looks to me a reasonable behavior because MQTT is over TCP (or TLS) but not http(s).
Allowing target group to be TCP does not mean you are exposing plaintext to the internet as long as the EMQX listener is SSL.

If you want encryption between NLB and EMQX, the NLB target group should be TLS but not TCP, not https either.
